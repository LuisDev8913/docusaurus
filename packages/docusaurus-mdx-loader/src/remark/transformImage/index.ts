/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {
  toMessageRelativeFilePath,
  posixPath,
  escapePath,
  getFileLoaderUtils,
} from '@docusaurus/utils';
import visit from 'unist-util-visit';
import path from 'path';
import url from 'url';
import fs from 'fs-extra';
import escapeHtml from 'escape-html';
import type {Plugin, Transformer} from 'unified';
import type {Image, Literal} from 'mdast';

const {
  loaders: {inlineMarkdownImageFileLoader},
} = getFileLoaderUtils();

interface PluginOptions {
  filePath: string;
  staticDirs: string[];
  siteDir: string;
}

function toImageRequireNode(node: Image, imagePath: string, filePath: string) {
  const jsxNode = node as Literal & Partial<Image>;
  let relativeImagePath = posixPath(
    path.relative(path.dirname(filePath), imagePath),
  );
  relativeImagePath = `./${relativeImagePath}`;

  const parsedUrl = url.parse(node.url);
  const hash = parsedUrl.hash ?? '';
  const search = parsedUrl.search ?? '';

  const alt = node.alt ? `alt={"${escapeHtml(node.alt)}"} ` : '';
  const src = `require("${inlineMarkdownImageFileLoader}${
    escapePath(relativeImagePath) + search
  }").default${hash ? ` + '${hash}'` : ''}`;
  const title = node.title ? ` title="${escapeHtml(node.title)}"` : '';

  Object.keys(jsxNode).forEach(
    (key) => delete jsxNode[key as keyof typeof jsxNode],
  );

  (jsxNode as Literal).type = 'jsx';
  jsxNode.value = `<img ${alt}src={${src}}${title} />`;
}

async function ensureImageFileExist(imagePath: string, sourceFilePath: string) {
  const imageExists = await fs.pathExists(imagePath);
  if (!imageExists) {
    throw new Error(
      `Image ${toMessageRelativeFilePath(
        imagePath,
      )} used in ${toMessageRelativeFilePath(sourceFilePath)} not found.`,
    );
  }
}

async function getImageAbsolutePath(
  imagePath: string,
  {siteDir, filePath, staticDirs}: PluginOptions,
) {
  if (imagePath.startsWith('@site/')) {
    const imageFilePath = path.join(siteDir, imagePath.replace('@site/', ''));
    await ensureImageFileExist(imageFilePath, filePath);
    return imageFilePath;
  } else if (path.isAbsolute(imagePath)) {
    // absolute paths are expected to exist in the static folder
    const possiblePaths = staticDirs.map((dir) => path.join(dir, imagePath));
    // eslint-disable-next-line no-restricted-syntax
    for (const possiblePath of possiblePaths) {
      const imageFilePath = possiblePath;
      if (await fs.pathExists(imageFilePath)) {
        return imageFilePath;
      }
    }
    throw new Error(
      `Image ${possiblePaths
        .map((p) => toMessageRelativeFilePath(p))
        .join(' or ')} used in ${toMessageRelativeFilePath(
        filePath,
      )} not found.`,
    );
  }
  // We try to convert image urls without protocol to images with require calls
  // going through webpack ensures that image assets exist at build time
  else {
    // relative paths are resolved against the source file's folder
    const imageFilePath = path.join(path.dirname(filePath), imagePath);
    await ensureImageFileExist(imageFilePath, filePath);
    return imageFilePath;
  }
}

async function processImageNode(node: Image, options: PluginOptions) {
  if (!node.url) {
    throw new Error(
      `Markdown image URL is mandatory in "${toMessageRelativeFilePath(
        options.filePath,
      )}" file`,
    );
  }

  const parsedUrl = url.parse(node.url);
  if (parsedUrl.protocol || !parsedUrl.pathname) {
    // pathname:// is an escape hatch,
    // in case user does not want his images to be converted to require calls going through webpack loader
    // we don't have to document this for now,
    // it's mostly to make next release less risky (2.0.0-alpha.59)
    if (parsedUrl.protocol === 'pathname:') {
      node.url = node.url.replace('pathname://', '');
    }
    return;
  }

  const imagePath = await getImageAbsolutePath(parsedUrl.pathname, options);
  toImageRequireNode(node, imagePath, options.filePath);
}

const plugin: Plugin<[PluginOptions]> = (options) => {
  const transformer: Transformer = async (root) => {
    const promises: Promise<void>[] = [];
    visit(root, 'image', (node: Image) => {
      promises.push(processImageNode(node, options));
    });
    await Promise.all(promises);
  };
  return transformer;
};

export default plugin;
