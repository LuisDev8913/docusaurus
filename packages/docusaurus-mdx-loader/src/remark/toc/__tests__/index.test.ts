/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import remark from 'remark';
import mdx from 'remark-mdx';
import vfile from 'to-vfile';
import plugin from '../index';
import headings from '../../headings/index';

const processFixture = async (name, options?) => {
  const filePath = path.join(__dirname, '__fixtures__', `${name}.md`);
  const file = await vfile.read(filePath);
  const result = await remark()
    .use(headings)
    .use(mdx)
    .use(plugin, options)
    .process(file);

  return result.toString();
};

test('non text phrasing content', async () => {
  const result = await processFixture('non-text-content');
  expect(result).toMatchSnapshot();
});

test('inline code should be escaped', async () => {
  const result = await processFixture('inline-code');
  expect(result).toMatchSnapshot();
});

test('text content', async () => {
  const result = await processFixture('just-content');
  expect(result).toMatchInlineSnapshot(`
    "export const toc = [
    	{
    		value: 'Endi',
    		id: 'endi',
    		level: 3
    	},
    	{
    		value: 'Endi',
    		id: 'endi-1',
    		level: 2
    	},
    	{
    		value: 'Yangshun',
    		id: 'yangshun',
    		level: 3
    	},
    	{
    		value: 'I ♥ unicode.',
    		id: 'i--unicode',
    		level: 2
    	}
    ];

    ### Endi

    \`\`\`md
    ## This is ignored
    \`\`\`

    ## Endi

    Lorem ipsum

    ### Yangshun

    Some content here

    ## I ♥ unicode.
    "
  `);
});

test('should export even with existing name', async () => {
  const result = await processFixture('name-exist');
  expect(result).toMatchInlineSnapshot(`
    "export const toc = [
    	{
    		value: 'Thanos',
    		id: 'thanos',
    		level: 2
    	},
    	{
    		value: 'Tony Stark',
    		id: 'tony-stark',
    		level: 2
    	},
    	{
    		value: 'Avengers',
    		id: 'avengers',
    		level: 3
    	}
    ];

    ## Thanos

    ## Tony Stark

    ### Avengers
    "
  `);
});

test('should export with custom name', async () => {
  const options = {
    name: 'customName',
  };
  const result = await processFixture('just-content', options);
  expect(result).toMatchInlineSnapshot(`
    "export const customName = [
    	{
    		value: 'Endi',
    		id: 'endi',
    		level: 3
    	},
    	{
    		value: 'Endi',
    		id: 'endi-1',
    		level: 2
    	},
    	{
    		value: 'Yangshun',
    		id: 'yangshun',
    		level: 3
    	},
    	{
    		value: 'I ♥ unicode.',
    		id: 'i--unicode',
    		level: 2
    	}
    ];

    ### Endi

    \`\`\`md
    ## This is ignored
    \`\`\`

    ## Endi

    Lorem ipsum

    ### Yangshun

    Some content here

    ## I ♥ unicode.
    "
  `);
});

test('should insert below imports', async () => {
  const result = await processFixture('insert-below-imports');
  expect(result).toMatchInlineSnapshot(`
    "import something from 'something';

    import somethingElse from 'something-else';

    export const toc = [
    	{
    		value: 'Title',
    		id: 'title',
    		level: 2
    	},
    	{
    		value: 'Test',
    		id: 'test',
    		level: 2
    	},
    	{
    		value: 'Again',
    		id: 'again',
    		level: 3
    	}
    ];

    ## Title

    ## Test

    ### Again

    Content.
    "
  `);
});

test('empty headings', async () => {
  const result = await processFixture('empty-headings');
  expect(result).toMatchInlineSnapshot(`
    "export const toc = [];

    # Ignore this

    ## 

    ## ![](an-image.svg)
    "
  `);
});
