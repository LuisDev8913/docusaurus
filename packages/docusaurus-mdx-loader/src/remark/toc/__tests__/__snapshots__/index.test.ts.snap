// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inline code should be escaped 1`] = `
"export const toc = [
	{
		value: '<code>&lt;Head /&gt;</code>',
		id: 'head-',
		level: 2
	},
	{
		value: '<code>&lt;Head&gt;Test&lt;/Head&gt;</code>',
		id: 'headtesthead',
		level: 3
	},
	{
		value: '<code>&lt;div /&gt;</code>',
		id: 'div-',
		level: 2
	},
	{
		value: '<code>&lt;div&gt; Test &lt;/div&gt;</code>',
		id: 'div-test-div',
		level: 2
	},
	{
		value: '<code>&lt;div&gt;&lt;i&gt;Test&lt;/i&gt;&lt;/div&gt;</code>',
		id: 'divitestidiv',
		level: 2
	}
];

## \`<Head />\`

### \`<Head>Test</Head>\`

## \`<div />\`

## \`<div> Test </div>\`

## \`<div><i>Test</i></div>\`
"
`;

exports[`non text phrasing content 1`] = `
"export const toc = [
	{
		value: '<em>Emphasis</em>',
		id: 'emphasis',
		level: 2
	},
	{
		value: '<strong>Importance</strong>',
		id: 'importance',
		level: 3
	},
	{
		value: '<del>Strikethrough</del>',
		id: 'strikethrough',
		level: 2
	},
	{
		value: '<i>HTML</i>',
		id: 'html',
		level: 2
	},
	{
		value: '<code>inline.code()</code>',
		id: 'inlinecode',
		level: 2
	}
];

## _Emphasis_

### **Importance**

## ~~Strikethrough~~

## <i>HTML</i>

## \`inline.code()\`
"
`;
